woocommerce/woocommerce
# Reduce duplicated code in Analytics controllers

_:warning: this PR targets a feature branch together with [multiple improvements to the Analytics codebase](https://github.com/woocommerce/woocommerce/pull/49425)_

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:




- [Reuse GenericController::prepare_item_for_response for `Reports\Controller` subclasses](https://github.com/woocommerce/woocommerce/commit/8bdccff32fa6d2d6c375a03ae7248bade75c2eb8) 
- [Reuse GenericController::get_collection_params for `Reports\Controller` subclasses.](https://github.com/woocommerce/woocommerce/commit/7b17eb3d1d8e75c95eb970fdbd44119bfbf86512) 
- [Move get_items code to shared GenericStatsController](https://github.com/woocommerce/woocommerce/commit/a2de7a8f6bbcdd49bca042a509d95eae38331a67)
- [Move shared paginable controlles get_items code to the generic class](https://github.com/woocommerce/woocommerce/pull/49427/commits/2bf60900b516cffa75d4c1bc1b837862b63a7b94)
- [Move fields param to GenericStatsController](https://github.com/woocommerce/woocommerce/pull/49427/commits/28dbf0bbdc2b209109ab3585bef7c8e33926172e)
- [Separate OrderAwareControllerTrait from ReportController](https://github.com/woocommerce/woocommerce/pull/49427/commits/07ed67381782997eb2d67a927340f475ba0f9fea) 
   to allow specific Analytics Controllers extends Generic(Stats)Controller directly, without extending the ReportController which is used to list reports.


Basically, it's meant not to change any behavior. However, due to unification, it did tweak a few things mostly from the developer perspective:
- Previously, only some controllers threw an error from `get_items` when the data was missing; some did not. Now controllers the following Controllers changed behavior to also for such an error:
	- Coupons
	- Customers
	- Downloads
	- Orders
	- Products
	- Taxes
- the error has one error code fro all controllers: `woocommerce_rest_reports_invalid_response` (previously there were report-specific)
   - `woocommerce_rest_reports_categories_invalid_response` -> `woocommerce_rest_reports_invalid_response`
- In Orders Controller the `$orders_data['order_number'], $orders_data['total_formatted']` are set in the `prepare_item_for_response` method not in the `get_items`So `prepare_item_for_response` function is expected to be called with bare data from Query
- In Products Controller extended-info is now sanitized in the `prepare_item_for_response` method, not in the `get_items`
- :warning: (breaking) Previosly some Controlers' `prepare_item_for_response` function was getting and expecting first argument to be casted to `object`, while the rest of controllers used data as returned from the Store. This PR unifies this behavior  for
   -  `Coupons\Stats\Controller`
   -  `Taxes\Controller`
   -  `Taxes\Stats\Controller`
   So, if you extend those classes and override `prepare_item_for_response` in a way that expects `object`, you will get an error. (I have not found such cases in the `all-plugins` repo.)
   All `woocommerce_rest_prepare_report_*` filters remain intact and inconsistent for backward compatibility.


There is another change that I extracted to track separately, as it may introduce more of a breaking change, but on the other hand, may be useful for some long term cleanups
https://github.com/woocommerce/woocommerce/pull/50002

### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR aims to change only the underlying code without changing any user-facing functionality, so there are no new things to test, only the regression.

1. Smoke test all Analytics pages, including `WooCommerce > Customers`
2. Check that the code of individual Controllers indeed gets simpler, shorter, and easier to maintain



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Reduce duplicated code in Analytics controllers
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
…troller` subclasses.
